src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -269,10 +347,36 @@ MergeTreeData::MutableDataPartPtr Fetcher::fetchPart(
     {
         {"endpoint",                getEndpointId(replica_path)},
         {"part",                    part_name},
-        {"client_protocol_version", toString(REPLICATION_PROTOCOL_VERSION_WITH_PARTS_UUID)},
+        {"client_protocol_version", toString(REPLICATION_PROTOCOL_VERSION_WITH_PARTS_S3_COPY)},
         {"compress",                "false"}
     });
 
+    if (try_use_s3_copy && disk_s3 && disk_s3->getType() != DB::DiskType::Type::S3)
```

Comment:
Let's move to separate function
`bool canFetchWithZeroCopyS3(bool try_use_s3_copy, disk_s3);`
And it will look like:
```
if (canFetchWithZeroCopyS3(try_use_s3_copy, disk_s3))
    uri.addQueryParameter("send_s3_metadata", "1");
```

Reply:
I think code it will be more complicated in that case, because code also fills 'disks_s3' array and can set 'try_use_s3_copy' to 'false', both variables used later.

Reply:
Ok